Chessington.GameEngine/Pieces/Pawn.cs

```diff
@@ -43,40 +45,60 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
                 availableLocations.Add(twoSquaresAhead);
             }
 
-            // Pawns can take pieces diagonally
-            var diagonalLeft = GetDiagonalMove(board, location, HoriztonalDirection.Left);
-            if (diagonalLeft != null) availableLocations.Add(diagonalLeft.Value);
-
-
-            var diagonalRight = GetDiagonalMove(board, location, HoriztonalDirection.Right);
-            if (diagonalRight != null) availableLocations.Add(diagonalRight.Value);
-
-            return availableLocations;
+            // Pawns can take pieces diagonally, and en passant
+            return availableLocations
+                .Concat(GetDiagonalMoves(board, location, -1))
+                .Concat(GetDiagonalMoves(board, location, 1));
         }
 
-        private Square? GetDiagonalMove(Board board, Square location, HoriztonalDirection horizontalDirection)
+        private IEnumerable<Square> GetDiagonalMoves(Board board, Square location, int horizontalDirection)
         {
+            var availableLocations = new List<Square>();
+
             var diagonal = new Square(
                 location.Row + verticalDirection,
                 location.Col + (int)horizontalDirection
             );
 
             if (Board.SquareIsOnBoard(diagonal)
-                && board.SquareIsOccupied(diagonal)
-                && board.GetPiece(diagonal).Player != Player)
+                   && board.SquareIsOccupied(diagonal)
+                   && board.GetPiece(diagonal).Player != Player)
             {
-                return diagonal;
+                availableLocations.Add(diagonal);
             }
-            else
+
+            // En passant
+            var side = new Square(location.Row, location.Col + horizontalDirection);
+            if (Board.SquareIsOnBoard(side) && !board.SquareIsOccupied(diagonal) &&
+                board.SquareIsOccupied(side) && board.GetPiece(side).CanBeEnPassantTaken(Player))
             {
-                return null;
+                availableLocations.Add(diagonal);
             }
+
+            return availableLocations;
         }
 
-        public new void MoveTo(Board board, Square newSquare)
+        public override void MoveTo(Board board, Square newSquare)
         {
+            hasJustDoubleMoved = Math.Abs(board.FindPiece(this).Row - newSquare.Row) == 2;
+
+            // En passant
+            var location = board.FindPiece(this);
+            var squareBehind = new Square(newSquare.Row - verticalDirection, newSquare.Col);
+            if (location.Col != newSquare.Col
+                && board.GetPiece(newSquare) == null
+                && board.SquareIsOccupied(squareBehind))
+            {
+                board.CapturePieceOnSquare(squareBehind);
+            }
+
             base.MoveTo(board, newSquare);
             hasMoved = true;
         }
+
+        public override bool CanBeEnPassantTaken(Player player)
```

Comment:
I would maybe name this `CanBeTakenEnPassant`? But not very important 🙂

Reply:
Yep sounds good. Changed in ba72ab6 🙂